agent/agent.go

```diff
@@ -1188,6 +1198,9 @@ func (a *Agent) JoinLAN(addrs []string) (n int, err error) {
 	a.logger.Printf("[INFO] agent: (LAN) joining: %v", addrs)
 	n, err = a.delegate.JoinLAN(addrs)
 	a.logger.Printf("[INFO] agent: (LAN) joined: %d Err: %v", n, err)
+	if err == nil && a.joinLANNotifier != nil {
```

Comment:
I never knew about this before, but there is a callback you can hook into to get notified when a server is known about, this might be a little later/safer time to notify since at that point you will know that a server is around to start to handle requests - https://github.com/hashicorp/consul/blob/master/agent/consul/client.go#L289-L292. There's a hook on here already so you'd need to chain to that one, and that will get called multiple times, so you'd need to deal with that, but it's an interesting place to trigger this. It also looks like this doesn't get hooked on servers, just clients.